parlai/nn/r3f.py

```diff
@@ -0,0 +1,190 @@
+# LICENSE file in the root directory of this source tree.
+
+"""
+Implements extentions to NN code for transformers.
+
+These include custom fine-tuning losses and the like that are training detail extentions
+rather than new model architectures in themselves.
+"""
+
+from contextlib import AbstractContextManager
+import re
+import torch
+import torch.nn.functional as F
+
+from parlai.core.params import ParlaiParser
+
+R3F_NOISE_NORMAL = "normal"
+R3F_NOISE_UNIFORM = "uniform"
+
+
+class R3FMixin(object):
+    """
+    Class that helps implement "Better Fine-Tuning by Reducing Representational
+    Collapse", Aghajanyan et al.
+
+    (2020). https://arxiv.org/ans/2008.03156.
+    """
+
+    @staticmethod
+    def add_cmdline_args(argparser: ParlaiParser):
+        group = argparser.add_argument_group('R3F fine-tuning Args')
+        group.add_argument('--r3f-eps', type=float, default=1e-5, help='noise eps')
+        group.add_argument(
+            '--r3f-lambda',
+            type=float,
+            default=1.0,
+            help='lambda for combining logistic loss and noisy KL loss',
+        )
+        group.add_argument(
+            '--r3f-noise-type',
+            type=str,
+            default=R3F_NOISE_UNIFORM,
+            choices=[R3F_NOISE_NORMAL, R3F_NOISE_UNIFORM],
+            help='type of noises for RXF methods',
+        )
+        group.add_argument(
+            '--r3f-encoder-noise',
+            type=bool,
+            default=False,
+            help='Add noise to encoder.',
+        )
+        group.add_argument(
+            '--r3f-decoder-noise',
+            type=bool,
+            default=False,
+            help='Add noise to decoder.',
+        )
+        return argparser
+
+    def set_r3f_settings_from_opts(self, opts):
+        self.r3f_eps = opts.get('r3f_eps')
+        self.r3f_noise_type = opts.get('r3f_noise_type')
+        if self.r3f_noise_type == R3F_NOISE_NORMAL:
+            self.r3f_noise_sampler = torch.distributions.normal.Normal(
+                loc=0.0, scale=self.r3f_eps
+            )
+        elif self.r3f_noise_type == R3F_NOISE_UNIFORM:
+            self.r3f_noise_sampler = torch.distributions.uniform.Uniform(
+                low=-self.r3f_eps, high=self.r3f_eps
+            )
+        self.r3f_lambda = opts.get("r3f_lambda")
+
+        self.noise_encoder = opts.get("r3f_encoder_noise")
+        self.noise_decoder = opts.get("r3f_decoder_noise")
+
+        # Find embedding values and store locally so we don't have to find them each turn
+        self.r3f_embeddings = {}
+        self.is_generative_model = True  # assume true until we find otherwise
+        self._find_embeddings(self.model)
+
+    def compute_loss(self, batch, return_output=False):
+        if not hasattr(self, 'r3f_lambda'):
+            self.set_r3f_settings_from_opts(self.opt)
+        loss, standard_output = super().compute_loss(batch, True)
+        with R3FNoiseEmbeddingContextManager(self, self.model) as r3f:
```

Comment:
love this final verison

Reply:
To be fair, this looks pretty much the same as what was initially planned out?

Reply:
👀